libs/agno/agno/models/base.py

```diff
@@ -145,6 +180,69 @@ def to_dict(self) -> Dict[str, Any]:
     def get_provider(self) -> str:
         return self.provider or self.name or self.__class__.__name__
 
+    def _filter_messages(self, messages: List[Message], num_tool_calls_to_keep: int) -> int:
+        """
+        Filter messages to keep only the most recent N tool calls.
+
+        Args:
+            messages: List of messages
+            num_tool_calls_to_keep: Number of recent tool calls to keep
+
+        Returns:
+            Number of tool calls filtered out
+        """
+        # Count total tool calls (not messages) - each tool result = 1 tool call
+        total_tool_calls = sum(1 for m in messages if m.role == "tool")
+
+        if total_tool_calls <= num_tool_calls_to_keep:
+            return 0
+
+        # Collect tool_call_ids to keep (most recent N)
+        tool_call_ids_to_keep = []
+        for msg in reversed(messages):
```

Comment:
You can probably use this for loop to already filter. 2 loops shouldn't be necessary I think

Reply:
So basically combine the 2 loops